pipeline/deploy-staging.sh

```diff
@@ -21,10 +35,30 @@ then
 fi
 
 git clone "$CNS_CSI_STAGING_REPO" || exit 1
+
+# Update the CNS-CSI manifest files to capture any changes.
+cp -R manifests/supervisorcluster/* staging-cd/ || exit 1
 cd staging-cd || exit 1
 
-# Patch the yaml file with the driver and syncer images from build job.
-yq -i '(.spec.template.spec.containers[0].image = env(VSPHERE_CSI_CONTROLLER_IMAGE)) | (.spec.template.spec.containers[1].image = env(VSPHERE_SYNCER_IMAGE))' staging/patch.yaml || exit 1
+# Patch the CSI controller patch yaml file with the driver and syncer images from build job.
+yq -i '(.spec.template.spec.containers[0].image = env(VSPHERE_CSI_CONTROLLER_IMAGE)) | (.spec.template.spec.containers[1].image = env(VSPHERE_SYNCER_IMAGE))' staging/csi-controller-patch.yaml || exit 1
```

Comment:
The `csi-controller-patch.yaml` has sidecar versions hard-coded. Let's say if there's a SV upgrade in staging environment that has newer sidecars, will those be patched back to the hard-coded values in `csi-controller-patch.yaml`, whenever CSI controller/syncer patching happens ?

Reply:
For now, we will have to set the sidecar images as per the Supervisor. Ideally, we should bump the sidecars to the latest, but that needs to be further analyzed as the CSI yamls are currently based on the Supervisor version. We should also analyze whether CSI yamls can be independent of the Supervisor versions.

Reply:
> We can keep fixed sidercar versions in CSI yamls corresponding to each Supervisor version.
We did not follow this so far. We have shipped Supervisors with different sidecars for 1.22. The CSI in staging environments are running sidecar images that do not match the sidecars in the CSI repo - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/manifests/supervisorcluster/1.22/cns-csi.yaml. For example, the repo says the attacher sidecar is `localhost:5000/vmware.io/csi-attacher:v4.0.0_vmware.1` whereas the actual attacher sidecar running in staging env is `localhost:5000/vmware.io/csi-attacher:v3.4.0_vmware.1`. So, we need to override this so that CSI can continue to run in the staging environment.

Reply:
I see, that makes sense.  
Btw do we know why do have such a disparity? Also if there's a Supervisor upgrade in staging environment that updates the sidecars, we will have to manually update those values in CSI controller patch file also.

Reply:
The sidecars are updated to solve the bugs related to CSI. So, we ended up shipping Supervisors with different sidecars for 1.22.

> we will have to manually update those values in CSI controller patch file also.

Yes, see https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/2297#discussion_r1145369869. We need to solve this problem, can we put this in the backlog?